File: interviewBit-golang/src/trees/problems/nextGreaterNumberBst.go
package trees

import (
	"interviewBit/src/utils"
)

func NextGreaterNumberBst(A *utils.TreeNode, B int) *utils.TreeNode {

	st := make([]*utils.TreeNode, 0)
	for len(st) > 0 || A != nil {
		for A != nil {
			st = append(st, A)
			A = A.Left
		}

		A = st[len(st)-1]
		st = st[:len(st)-1]

		if A.Value > B {
			return A
		}
		A = A.Right
	}

	return nil
}

File: interviewBit-cpp/src/trees/next-greater-number-bst.cpp
#include <iostream>
#include "trees.h"

using namespace std;

// Optimised Approach 1:
// T(n) : O(n) ; S(n) : O(1)
TreeNode* solveNextGreaterNumberBst(TreeNode* A, int B) {

    TreeNode *nextGreaterBstNode = nullptr;
    while (A != nullptr) {
        if (A->val <= B) {
            A = A->right;
        } else {
            nextGreaterBstNode = A;
            A = A->left;
        }
    }
    
    return nextGreaterBstNode;
}

// Driver Code for testing
int main() {

    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    
    return 0;
}